timeutil/schedule.go

```diff
@@ -0,0 +1,251 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package timeutil
+
+import (
+	"fmt"
+	"math/rand"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var validTime = regexp.MustCompile(`^([0-9]|0[0-9]|1[0-9]|2[0-3]):([0-5][0-9]):?([0-5][0-9])?$`)
+
+type TimeOfDay struct {
+	Hour   int
+	Minute int
+	Second int
+}
+
+func (td TimeOfDay) Less(other TimeOfDay) bool {
+	if td.Hour < other.Hour {
+		return true
+	}
+	if td.Minute < other.Minute {
+		return true
+	}
+	if td.Second < other.Second {
+		return true
+	}
+	return false
+}
+
+// ParseTime parses a string that contains hour:minute and returns
+// an TimeOfDay type or an error
+func ParseTime(s string) (t TimeOfDay, err error) {
+	m := validTime.FindStringSubmatch(s)
+	if len(m) == 0 {
+		return t, fmt.Errorf("cannot parse %q", s)
+	}
+	t.Hour, err = strconv.Atoi(m[1])
+	if err != nil {
+		return t, fmt.Errorf("cannot parse %q: %s", m[1], err)
+	}
+	t.Minute, err = strconv.Atoi(m[2])
+	if err != nil {
+		return t, fmt.Errorf("cannot parse %q: %s", m[2], err)
+	}
+	if m[3] != "" {
+		t.Second, err = strconv.Atoi(m[3])
+		if err != nil {
+			return t, fmt.Errorf("cannot parse %q: %s", m[3], err)
+		}
+	}
+	return t, nil
+}
+
+// Schedule defines a start and end time and an optional weekday in which
+// events should run.
+type Schedule struct {
+	Start TimeOfDay
+	End   TimeOfDay
+
+	Weekday string
+}
+
+func (sched *Schedule) Next(last time.Time) (start, end time.Time) {
+	now := timeNow()
+	wd := time.Weekday(weekdayMap[sched.Weekday])
+
+	t := last
+	for {
+		// slightly stupid, we move only a single day forward
+		t = t.AddDate(0, 0, 1)
+
+		a := time.Date(t.Year(), t.Month(), t.Day(), sched.Start.Hour, sched.Start.Minute, sched.Start.Second, 0, time.Local)
+		b := time.Date(t.Year(), t.Month(), t.Day(), sched.End.Hour, sched.End.Minute, sched.End.Second, 0, time.Local)
+
+		// we have not hit the right day yet
+		if sched.Weekday != "" && a.Weekday() != wd {
+			continue
+		}
+		// schedule is right now
+		if a.Before(now) && b.After(now) {
+			return a, b
+		}
+		// not yet after now
+		if !a.After(now) {
+			continue
+		}
+
+		return a, b
+	}
+}
+
+func randDur(dur time.Duration) time.Duration {
+	return time.Duration(rand.Int63n(int64(dur)))
+}
+
+var (
+	timeNow     = time.Now
+	maxDuration = 14 * 24 * time.Duration(time.Hour)
+)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+// Next will return the duration until a random time in the next
+// schedule window.
+func Next(schedule []*Schedule, last time.Time) time.Duration {
+	now := timeNow()
+
+	// delay more than 14 days, always update
+	if now.Sub(last) > maxDuration {
+		return 0
+	}
+
+	var a, b time.Time
+	for _, sched := range schedule {
+		start, end := sched.Next(last)
+		if a.IsZero() || start.After(now) && start.Before(a) {
```

Comment:
Why is this checking that `start.After(now)`?  This means we'd never pick an interval which we're already inside of.

Assuming the above suggestion, this might end up just as `if start.Before(a) {`.